tests/network/remote/remotes.c

```diff
@@ -28,30 +28,78 @@ void test_network_remote_remotes__cleanup(void)
 
 void test_network_remote_remotes__parsing(void)
 {
+	char *url = NULL;
 	git_remote *_remote2 = NULL;
 
 	cl_assert_equal_s(git_remote_name(_remote), "test");
 	cl_assert_equal_s(git_remote_url(_remote), "git://github.com/libgit2/libgit2");
 	cl_assert(git_remote_pushurl(_remote) == NULL);
 
-	cl_assert_equal_s(git_remote__urlfordirection(_remote, GIT_DIRECTION_FETCH),
-					  "git://github.com/libgit2/libgit2");
-	cl_assert_equal_s(git_remote__urlfordirection(_remote, GIT_DIRECTION_PUSH),
-					  "git://github.com/libgit2/libgit2");
+	cl_git_pass(git_remote__urlfordirection(_remote, GIT_DIRECTION_FETCH, NULL, &url));
+	cl_assert_equal_s(url, "git://github.com/libgit2/libgit2");
+
+	free(url);
+
+	cl_git_pass(git_remote__urlfordirection(_remote, GIT_DIRECTION_PUSH, NULL, &url));
+	cl_assert_equal_s(url, "git://github.com/libgit2/libgit2");
+
+	free(url);
 
 	cl_git_pass(git_remote_lookup(&_remote2, _repo, "test_with_pushurl"));
 	cl_assert_equal_s(git_remote_name(_remote2), "test_with_pushurl");
 	cl_assert_equal_s(git_remote_url(_remote2), "git://github.com/libgit2/fetchlibgit2");
 	cl_assert_equal_s(git_remote_pushurl(_remote2), "git://github.com/libgit2/pushlibgit2");
 
-	cl_assert_equal_s(git_remote__urlfordirection(_remote2, GIT_DIRECTION_FETCH),
-					  "git://github.com/libgit2/fetchlibgit2");
-	cl_assert_equal_s(git_remote__urlfordirection(_remote2, GIT_DIRECTION_PUSH),
-					  "git://github.com/libgit2/pushlibgit2");
+	cl_git_pass(git_remote__urlfordirection(_remote2, GIT_DIRECTION_FETCH, NULL, &url));
+	cl_assert_equal_s(url, "git://github.com/libgit2/fetchlibgit2");
+
+	free(url);
+
+	cl_git_pass(git_remote__urlfordirection(_remote2, GIT_DIRECTION_PUSH, NULL, &url));
+	cl_assert_equal_s(url, "git://github.com/libgit2/pushlibgit2");
+
+	free(url);
 
 	git_remote_free(_remote2);
 }
 
+static int test_network_remote_remotes__urlresolve_callback(const char *url, char **url_resolved, int direction, void *payload)
+{
+	const char *url_out;
+
+	cl_assert(git__strcmp(url, "git://github.com/libgit2/libgit2") == 0);
+	cl_assert(git__strcmp(payload, "payload") == 0);
+
+	url_out = (direction == GIT_DIRECTION_PUSH) ? "pushresolve" : "fetchresolve";
+
+	*url_resolved = git__strdup(url_out);
```

Comment:
I don't know how the custom allocators should be correctly used, I just try to match surrounding code, and since everywhint is using the `git__` style, I used that. So no `git__` allocators in test code?

Reply:
Well, that's a pretty specific case here: usually you'd be right, but outside users won't be able to access our `git__strdup`/underlying allocator, and instead will pass their own allocated pointer. This might not be fatal, but the Win32 build can use a different allocator for the library (or the user code might have its own), and it will complain.

Now, this pointer has to be freed *at some point*, but you can't before you've passed it to the library, and we don't give back control afterward to free it. Maybe there's a way to do that by passing a `git_buf` instead of a `void *`, but it needs some more thought because that free *has* to happen. And I might also be completely wrong about all of it, BTW 😉.

Reply:
I have no idea what the issue is and what you expect me to do here. In any case, it is a test case. What does it matter?

Reply:
Also, don't people pay M$ to fix their toy os? Why does Win32 require some custom allocator implementation instead of the POSIX ones? But probably off topic...

Reply:
Also, If users allocating stuff seems to be a problem, it seems to me there is a bigger underlying issue here.

Reply:
@tiennou I've replaced the `**char` with a buffer

Reply:
Toy OS or not 😒, there's nothing to "fix".  You have to use a `free` function that your allocator specifies.  If you're using `malloc`, then you need to call `free`.  If you _stop_ using `malloc`, and you pick some _other_ allocator, like `jemalloc` or hoard or `tcmalloc`, then you have to use _it's_ free function.

If you take some piece of memory that was given to you by `tcmalloc`, and hand it to libgit2 and libgit2 naively calls `free` on it, then you will crash.

This is possible even within libgit2 itself:  libgit2 has a pluggable allocator, so that if you want it to use `jemalloc`, you can.  So `git__malloc` might not be `malloc`.  So you need to make sure to use `git__free` instead of `free`.

Now, why does this make a difference for Windows?  Because there's a debugging allocator that's frequently used since it adds guard pages and the like, and makes debugging memory issues easier in your application.  But you can't mix and match: if some application is built with the debugging allocator, and allocates some memory, and libgit2 is _not_ built with the debugging allocator, and tries to free that memory, then the application will crash.

But this is not unique to Windows.  This is just a particular instance of the larger problem, on all allocators.  Whoever creates the memory also needs to free the memory.

As a result, libgit2 never calls `free` on anything that the user gave us.  And we never return something and expect a user to `free` it.

Reply:
> Whoever creates the memory also needs to free the memory.

Because those allocators might be different, thanks for the clarification @ethomson.

In short, one of the hacky solutions (which I'm not actually recommending but which is somewhat sane) to fix this issue is to use a `static char *`, free it at the beginning of the function, alloc it if it's needed, all because libgit2 doesn't know how the free should be done. You lose reentrancy by doing that, but it works.

What need to happen with that memory must be closely checked against what kind of memory-borrowing we provide, and I'm not sure we have support for passing around a `git_buf` with a custom deallocator… I'll try to review our options tonight, if that's okay.

Reply:
You could use ```alloca``` (```_alloca``` on Windows) if the scope is limited to the function.

Reply:
@ethomson sorry for blowing up like this... feel a little guilty. It's just that this PR already takes so long, and then after many changes it seems as it would be discarded in the last minute... lost my cool for a bit.

Reply:
No need to apologize.  😀 

I'm sorry that this is taking so long, and I understand your frustration.  I regret that the core contributors have so little time to devote to the project right now.  So it will feel like we are silent, then swoop in only to disappear again for a long stretch.  I know from experience that this is both disheartening and frustrating as a contributor.  So, my apologies for that.  🙏

Reply:
I'm happy with this `git_buf` approach.  Callers can set it if they wish using `git_buf_set`, at which point we do a copy.

Reply:
Sorry for the hard feelings @eaigner. I was feeling pretty good about your PR, until I noticed the allocator problem, I wasn't that available yesterday, and it went downhill from there.

Right now I've squashed, polished, and refiled your PR as #5062. I don't really care which goes it, it's about the same code, I've just made some slight code moves for cleanliness + the struct upgrader (Note that I seem to have fixup-ed the wrong commit though, so it not quite how I wanted it to look: the upgrader changes are supposed to be on their own).

> Callers can set it if they wish using `git_buf_set`, at which point we do a copy.

~Isn't that private ? Right now the issue is that our private API is visible in the test suite, which is not true in general.~ My word, it is 😭.

Reply:
> Isn't that private ?

Right, it's not and historically for the same reason as here, to allow callers to give us data.  I admit it's a bit wonky but likely the least bad option afforded to us.  (Definitely better than the odb backend allocator, sigh.)

> I don't really care which goes...

I do; I want to merge this PR.  @eaigner has been exceedingly patient and made a lot of iterations on this PR and I want to see this one get merged.

(Putting myself in his shoes, if I'd done a lot of work and the result was a different PR getting merged then I'd be upset.)

Reply:
i don't mind. I just want to stop worrying about this PR, and start worrying about the next PR #5057

Reply:
> (Putting myself in his shoes, if I'd done a lot of work and the result was a different PR getting merged then I'd be upset.)

I do really want to thank you for that work — even though it looks like I rewrote half of it, it wouldn't have been written it in the first place if it had only been up to me 😁.

On the way forward, I don't mind taking care of the last mile it has to go through to land, but I'd prefer to do so from #5062.  The other solution is that I repurpose #5062 to be just the struct upgrading, then you can make use of that here to fix the bc-break ?

Reply:
Go ahead with #5062